### PR TITLE
create_project()

### PR DESCRIPTION
**Pull request by JT-39**
# Brief overview of changes

Two main files have been added to create the new function `create_project()`.
- `R/create_project.R`
- `inst/rstudio/templates/project/create_project.dcf`

The first includes the function to create and populate a new R project. It includes several files and folders by default such as `_analysis/analysis.qmd`, a `renv.lock` and `tests/`. There are options that allow one to alter the structure of the project (for example for package development or a publication pipeline).

The second file provides the binding to create a GUI option to create the R project. This appear when one clicks the "New project" at the top right of the RStudio IDE.

## Why are these changes being made?

These changes are being made as the solution for the issue - Pre-populated project template #70

## Detailed description of changes

- `R/create_project.R` added which is the main function to create the pre-populated project.
- `inst/rstudio/templates/project/create_project.dcf` which builds the GUI to start the project.
- Frederick logo added to above folder so it appears in the GUI.
- Amended the Imports and Suggest in the `DESCRIPTION` file to facilitate `create_project()`.
- Test of basic functionality - does it create the folders and files we expect.
- Test to see if the warning appears if `{renv}` is not installed.
- `devtools::check()` has been re-run, creating the `create_project()` vignette.
- The _pkgdown.yml has also been updated to include the new function.

## Additional information for reviewers

A set of tests have been added, however these are not exhaustive and could include more.

Currently the `init_renv` parameter of the function is set to TRUE as default. This initiates `{renv}` in the created project, which is best practice. However, the `renv.lock` is setup to record several packages (`{rmarkdown}`, `{testthat}`, `{renv}`, `{stringr}`). There is a minor issue that these packages are not installed in the new project and so the message appears:

`"- None of the packages recorded in the lockfile are currently installed."`

This can be rectified by executing `renv::restore()` which will install those packages. These packages are needed to create the tests and run the Quarto file. They could be reviewed. Potentially the `init_renv` is set to `FALSE` so it is more optional but I think promoting `{renv}` is best practice.

More documentation will be necessary to describe and explain how the function work, especially the above.

The actual structure of the project created should also be reviewed, especially for the publication project pipeline option (I have just guessed this and assume there is an existing project template which would go better there).

## Issue ticket number/s and link

Pre-populated project template #70

### CODE REVIEW

File: R/create_project.R
Comment (chfoster) [inline R/create_project.R:8]: These are Boolean as well I think & needed for it to run, just like init_renv

File: R/create_project.R
Comment (chfoster) [inline R/create_project.R:62]: I'd be tempted to call this functions.R for consistency, @rmbielby opinions?

**Comment by chfoster** [comment]
Don't think this error message is necessarily right: 
image
 - my REPOS folder isn't a project, and also why 3 options but 1 and 3 both mean no? 😂

File: R/create_project.R
Comment (cjrace) [inline R/create_project.R:255]: It could be nice to add a message at the end of the function confirming the project has been made and telling the user to go to that directory?

**Review by cjrace — CHANGES_REQUESTED** [review]
Hey @JT-39 - thanks for raising this, this is absolutely fantastic! Love the GUI part of it!

Have left a few specific comments on the code, and have a couple of more general thoughts too:

1. Is there a main analysis script that you're expecting people will run? (thinking of something like a run_all.R in the root or R directories)

2. This could be a prime time to look at adding the praise or other equivalent package from #61, could add something fun into the Rprofile as a DfE specific greeting

3. Have you considered a separate vignette to walk through how to download the package and then use the GUI to create a new project? Thinking of taking it right to the very basics for someone still unfamiliar with R. This could be done after this PR if too much work to add here.

**Comment by cjrace** [comment]
[Jake Tufts PR review](https://trello.com/c/X5EQywS5/1439-jake-tufts-pr-review)

File: R/create_project.R
Comment (cjrace) [reply R/create_project.R:62]: Had a chat again with the wider team to try and get some consistency here with how we do shiny apps too and we've come up with `helper_functions.R` as the name for this kind of script - happy to go with this @JT-39?

**Comment by JT-39** [comment]
> Don't think this error message is necessarily right: image
> 
> * my REPOS folder isn't a project, and also why 3 options but 1 and 3 both mean no? 😂

I don't seem to get that error message. I created a new folder from the `C:\` directory that match you folder names, and it worked fine with no message. Is there is a .git file hidden in your `C:\REPOS\` folder?
Also I think the two options being the same is a feature from `{usethis}` and is so the user does not just choose an option accidentally and reads them with care before making any changes.

File: R/create_project.R
Comment (JT-39) [reply R/create_project.R:8]: Yep, good spot, will change these. I have set `include_structure_for_pkg()` and `create_publication_proj()` to default as `FALSE` but left `include_github_gitignore()` as open, as think it would be good practice for the user to decide this?

File: R/create_project.R
Comment (JT-39) [reply R/create_project.R:62]: Yeah that is certainly fine with me. Tbh, I do like `utils.R` which is what I have found myself using more and may also be more generalisable to the wider R community. But `helper_functions.R` is perhaps a better description of the script and its content?

File: .Rbuildignore
Comment (JT-39) [reply .Rbuildignore:1]: Yeah good point. `{renv}` was not playing ball with `{covr}`. I am not sure if it is best practice to include `{renv}` in a package or not, I have found support for both sides. This is a link to the issue:
https://github.com/r-lib/covr/issues/571

I will remove those lines for the meantime though.

File: tests/testthat/test-create_project.R
Comment (JT-39) [reply tests/testthat/test-create_project.R:2]: Yep this has been done. I've used `suppressMessages()` around the `create_project()` call in test script.

I wonder though if you would also be against this nosiness when a user uses the function to create a project, as it will output those messages (due to this being what the `{usethis}` functions like `usethis::use_testthat()` outputs).

File: R/create_project.R
Comment (JT-39) [reply R/create_project.R:36]: Yep, done.

File: R/create_project.R
Comment (JT-39) [reply R/create_project.R:50]: Yep agreed, will change these. I have set include_structure_for_pkg() and create_publication_proj() to default as FALSE but left include_github_gitignore() as open, as think it would be good practice for the user to decide this?

Will include a `stop()` to check all params are specified.

File: R/create_project.R
Comment (JT-39) [reply R/create_project.R:255]: I like this idea. Have implemented with this:

```
cat(
    paste0("Your new dfeR project has been successfuly created! ",
           emo::ji("mortar_board"), "\n",
           "It exists at the file path: '", path, "'")
  )
```

May need to discuss nosiness of output as shows all the files created and the renv packages.

**Comment by JT-39** [comment]
> Hey @JT-39 - thanks for raising this, this is absolutely fantastic! Love the GUI part of it!
> 
> Have left a few specific comments on the code, and have a couple of more general thoughts too:
> 
> 1. Is there a main analysis script that you're expecting people will run? (thinking of something like a run_all.R in the root or R directories)
> 2. This could be a prime time to look at adding the praise or other equivalent package from [Find a way to make use of the `praise` package #61](https://github.com/dfe-analytical-services/dfeR/issues/61), could add something fun into the Rprofile as a DfE specific greeting
> 3. Have you considered a separate vignette to walk through how to download the package and then use the GUI to create a new project? Thinking of taking it right to the very basics for someone still unfamiliar with R. This could be done after this PR if too much work to add here.

Thanks! To answer your points:

1. Not really at the moment. I think there should be, but unsure what to call it/how it fits in with the `02_analysis/` folder. Does it replace it? Should it be a `.qmd` or just plain `.R` script. Name it `main`?

2. I like this idea! I've added a .Rprofile and used `.First <- function() {message('Welcome to your dfeR project!')}`. I tried adding {praise} but it didn't like the package coming from git with `{renv}`. Also, when entering the project for first time the user is a bit blasted from `{renv}` about installing packages from the renv.lock file. Should discuss.

3.  Yes, I will do this. Another good idea. Will explore how to do this!

File: DESCRIPTION
Comment (JT-39) [reply DESCRIPTION:32]: Yep, removed and is all fine.

**Comment by cjrace** [comment]
All looks good @JT-39, happy for this to get merged in once a couple of small last bits have been tidied:

1. Now you've added emo, the package won't install for the first time as it can't find it. Good news is I think this is easy to fix, as emo is a GitHub repo you need to add a new line with `Remotes: hadley/emo` into the DESCRIPTION file so that it knows to look at GitHub to install it's dependencies too. 
2. You should add yourself as a contributor in the DESCRIPTION 🎉 copy how Jen did her line if you want an example of how to do it
3. I'd recommend running `usethis::use_tidy_description()` after 1. and 2. to reorder and format the DESCRIPTION file neatly too
4. There's conflicts in the inst/WORDLIST file, from a glance it looks like you've added 'center' in but the way it remakes the list has reordered it and caused the conflict. Assuming you'll be able to resolve this pretty easily / quickly. Resolving this will also then unblock all the CI checks on this PR (I've ran checks locally and it seems to pass fine so not expecting any issues with those)
5. As this is pretty much ready to go, do you want to have a go at bumping the package up a minor version as this is a nice new feature. So make it v0.4.0 using the notes in the contributing file for how to do this with `usethis` and add a quick note in the space it'll give you in the NEWS.md file about the new function.

Once the above is done, assuming the CI checks all pass I'll approve, and once merged I'll release the GitHub version to match.

I've also raised #76 to cover the vignette separately, if that goes in before anything else, it can go in as a patch update so would then be v0.4.1 (though we can agree this on the PR at the time it's ready)

File: R/create_project.R
Comment (cjrace) [reply R/create_project.R]: Yeah you can just do that without doing library if the package is installed. Either way works though and I suspect you're right that not too many would know the `package::` syntax so marking this as resolved

**Comment by JT-39** [comment]
> All looks good @JT-39, happy for this to get merged in once a couple of small last bits have been tidied:
> 
> 1. Now you've added emo, the package won't install for the first time as it can't find it. Good news is I think this is easy to fix, as emo is a GitHub repo you need to add a new line with `Remotes: hadley/emo` into the DESCRIPTION file so that it knows to look at GitHub to install it's dependencies too.
> 2. You should add yourself as a contributor in the DESCRIPTION 🎉 copy how Jen did her line if you want an example of how to do it
> 3. I'd recommend running `usethis::use_tidy_description()` after 1. and 2. to reorder and format the DESCRIPTION file neatly too
> 4. There's conflicts in the inst/WORDLIST file, from a glance it looks like you've added 'center' in but the way it remakes the list has reordered it and caused the conflict. Assuming you'll be able to resolve this pretty easily / quickly. Resolving this will also then unblock all the CI checks on this PR (I've ran checks locally and it seems to pass fine so not expecting any issues with those)
> 5. As this is pretty much ready to go, do you want to have a go at bumping the package up a minor version as this is a nice new feature. So make it v0.4.0 using the notes in the contributing file for how to do this with `usethis` and add a quick note in the space it'll give you in the NEWS.md file about the new function.
> 
> Once the above is done, assuming the CI checks all pass I'll approve, and once merged I'll release the GitHub version to match.
> 
> I've also raised #76 to cover the vignette separately, if that goes in before anything else, it can go in as a patch update so would then be v0.4.1 (though we can agree this on the PR at the time it's ready)


Think this has all been done now! 😀

There are conflicts but from the update of the DESCRIPTION and NEWS file due to the increment in version, assume this is fine?

**Review by cjrace — APPROVED** [review]
Happy to approve! Raised #77 to look more into the linting as it shouldn't affect usage so we can deal with it separately

**Pull request merged**
